firmware/defmt-rtt/src/lib.rs

```diff
@@ -39,76 +44,186 @@
+            cs_restore: UnsafeCell::new(critical_section::RestoreState::invalid()),
+            encoder: UnsafeCell::new(defmt::Encoder::new()),
+        }
+    }
+
+    /// Acquire the defmt encoder.
+    fn acquire(&self) {
         // safety: Must be paired with corresponding call to release(), see below
         let restore = unsafe { critical_section::acquire() };
 
-        // safety: accessing the atomic without CAS is OK because we have acquired a critical section.
-        if TAKEN.load(Ordering::Relaxed) {
+        // NB: You can re-enter critical sections but we need to make sure
+        // no-one does that.
+        if self.taken.load(Ordering::Relaxed) {
             panic!("defmt logger taken reentrantly")
         }
 
-        // safety: accessing the atomic without CAS is OK because we have acquired a critical section.
-        TAKEN.store(true, Ordering::Relaxed);
-
-        // safety: accessing the `static mut` is OK because we have acquired a critical section.
-        unsafe { CS_RESTORE = restore };
+        // no need for CAS because we are in a critical section
+        self.taken.store(true, Ordering::Relaxed);
 
-        // safety: accessing the `static mut` is OK because we have acquired a critical section.
+        // safety: accessing the cell is OK because we have acquired a critical
+        // section.
         unsafe {
-            let encoder: &mut defmt::Encoder = &mut *core::ptr::addr_of_mut!(ENCODER);
-            encoder.start_frame(do_write)
+            self.cs_restore.get().write(restore);
+            let encoder: &mut defmt::Encoder = &mut *self.encoder.get();
+            encoder.start_frame(|b| {
+                _SEGGER_RTT.up_channel.write_all(b);
+            });
         }
     }
 
-    unsafe fn flush() {
-        // safety: accessing the `&'static _` is OK because we have acquired a critical section.
-        handle().flush();
-    }
+    /// Write bytes to the defmt encoder.
+    unsafe fn write(&self, bytes: &[u8]) {
+        if !self.taken.load(Ordering::Relaxed) {
+            panic!("defmt write out of context")
+        }
 
-    unsafe fn release() {
-        // safety: accessing the `static mut` is OK because we have acquired a critical section.
+        // safety: accessing the cell is OK because we have acquired a critical
+        // section.
         unsafe {
-            let encoder: &mut defmt::Encoder = &mut *core::ptr::addr_of_mut!(ENCODER);
-            encoder.end_frame(do_write);
+            let encoder: &mut defmt::Encoder = &mut *self.encoder.get();
+            encoder.write(bytes, |b| {
+                _SEGGER_RTT.up_channel.write_all(b);
+            });
         }
+    }
 
-        // safety: accessing the atomic without CAS is OK because we have acquired a critical section.
-        TAKEN.store(false, Ordering::Relaxed);
+    /// Flush the encoder
+    unsafe fn flush(&self) {
+        if !self.taken.load(Ordering::Relaxed) {
```

Comment:
I don't think relaxed is correct here? Nothing here enforces an ordering relationship for this load if I see that right

Reply:
Ah wait nevermind, that's the unsafe part, we expect the lock to be acquired here I suppose

Reply:
👍